File: exam-3/PlaylistManager-Python/Solution.py
class Song:
    def __init__(self,title,artist,album,genre,duration):
        self.title = title
        self.artist = artist
        self.album = album
        self.genre = genre
        self.duration  = duration
        self.list = []

    def display_details(self):
        return f"Title: {self.title}, Artist: {self.artist}, Album: {self.album}, Genre: {self.genre}, Duration: {self.duration}"
    
class Playlist:
    def __init__(self,name):
        self.name = name
        self.list = []

    def add_song(self,song):
        self.list.append(song)
        # print("b",self.list)
        
    def remove_song(self,song):
        for i in self.list:
            if i.title == song:
                self.list.remove(i)
                # print("a",self.list)
                return True
        return False
    def get_songs(self):
        return  self.list
    
    def filter_songs(self,criteria,value):
        filtered = []
        if criteria == "artist":
            for i in self.list:
                if i.artist == value:
                    filtered.append(i)
            return filtered
        
        if criteria == "genre":
            for i in self.list:
                if i.genre == value:
                    filtered.append(i)
            return filtered
        
        
        if criteria == "title":
            for i in self.list:
                if i.title == value:
                    filtered.append(i)
            return filtered
        

        if criteria == "album":
            for i in self.list:
                if i.album == value:
                    filtered.append(i)
            return filtered
        

                    
            
    def search_songs(self,keyword):
        key = []
        for i in self.list:
            if keyword.lower() in i.title.lower() or keyword.lower() in i.artist.lower() or keyword.lower() in  i.album.lower() or keyword.lower() in  i.genre.lower():
                key.append(i)
        # print(key)
        return key
    
    
class PlaylistManager:
    def __init__(self):
        self.palylists = []

    def create_playlist(self,name):
        self.palylists.append(Playlist(name))
            
            # print(self.palylists)
        
    def delete_playlist(self,name):
        for i in self.palylists:
            if i.name == name:
                self.palylists.remove(i)
                return True
        return False
    
    def get_playlist(self,name):
        # print(len(self.palylists))
        for i in self.palylists:
            if i.name == name:
                return i
        return None
    
    def list_playlists(self):
        return self.palylists
    
    def cross_playlist_search(self,keyword):
        songs = []
        # print(keyword)

        for i in self.palylists:
            s = i.search_songs(keyword)
            # print(s)
            songs.extend(s)
            # print(songs)
        return songs
    

File: exam-3/PlaylistManager-Python/Solution.java
import java.util.ArrayList;
import java.util.List;

class Song {
    String title;
    String artist;
    String album;
    String genre;
    String duration;
    List<Song> list;

    public Song(String title, String artist, String album, String genre, String duration) {
        this.title = title;
        this.artist = artist;
        this.album = album;
        this.genre = genre;
        this.duration = duration;
        this.list = new ArrayList<>();
    }

    public String displayDetails() {
        return "Title: " + this.title + ", Artist: " + this.artist + ", Album: " + this.album + ", Genre: " + this.genre + ", Duration: " + this.duration;
    }
}

class Playlist {
    String name;
    List<Song> list;

    public Playlist(String name) {
        this.name = name;
        this.list = new ArrayList<>();
    }

    public void addSong(Song song) {
        this.list.add(song);
    }

    public boolean removeSong(String songTitle) {
        for (Song song : this.list) {
            if (song.title.equals(songTitle)) {
                this.list.remove(song);
                return true;
            }
        }
        return false;
    }

    public List<Song> getSongs() {
        return this.list;
    }

    public List<Song> filterSongs(String criteria, String value) {
        List<Song> filtered = new ArrayList<>();
        for (Song song : this.list) {
            switch (criteria) {
                case "artist":
                    if (song.artist.equals(value)) {
                        filtered.add(song);
                    }
                    break;
                case "genre":
                    if (song.genre.equals(value)) {
                        filtered.add(song);
                    }
                    break;
                case "title":
                    if (song.title.equals(value)) {
                        filtered.add(song);
                    }
                    break;
                case "album":
                    if (song.album.equals(value)) {
                        filtered.add(song);
                    }
                    break;
            }
        }
        return filtered;
    }

    public List<Song> searchSongs(String keyword) {
        List<Song> key = new ArrayList<>();
        for (Song song : this.list) {
            if (song.title.toLowerCase().contains(keyword.toLowerCase()) ||
                song.artist.toLowerCase().contains(keyword.toLowerCase()) ||
                song.album.toLowerCase().contains(keyword.toLowerCase()) ||
                song.genre.toLowerCase().contains(keyword.toLowerCase())) {
                key.add(song);
            }
        }
        return key;
    }
}

class PlaylistManager {
    List<Playlist> playlists;

    public PlaylistManager() {
        this.playlists = new ArrayList<>();
    }

    public void createPlaylist(String name) {
        this.playlists.add(new Playlist(name));
    }

    public boolean deletePlaylist(String name) {
        for (Playlist playlist : this.playlists) {
            if (playlist.name.equals(name)) {
                this.playlists.remove(playlist);
                return true;
            }
        }
        return false;
    }

    public Playlist getPlaylist(String name) {
        for (Playlist playlist : this.playlists) {
            if (playlist.name.equals(name)) {
                return playlist;
            }
        }
        return null;
    }

    public List<Playlist> listPlaylists() {
        return this.playlists;
    }

    public List<Song> crossPlaylistSearch(String keyword) {
        List<Song> songs = new ArrayList<>();
        for (Playlist playlist : this.playlists) {
            List<Song> s = playlist.searchSongs(keyword);
            songs.addAll(s);
        }
        return songs;
    }
}

